fuzzers/frida_executable_libpng/src/lib.rs

```diff
@@ -0,0 +1,22 @@
+use std::process::exit;
+
+mod fuzzer;
+
+#[no_mangle]
+fn __libc_csu_fini() {}
+
+#[no_mangle]
+fn __libc_csu_init() {}
+
+#[no_mangle]
+pub unsafe extern "C" fn __libc_start_main(
+    main: extern "C" fn(isize, *const *const u8) -> isize,
+    _argc: isize,
+    _argv: *const *const char,
+) {
+    unsafe {
+        fuzzer::lib(main);
+    }
+
```

Comment:
I think it's safer to call the original __libc_start_main here.

can you try something like this?
https://gist.github.com/apsun/1e144bf7639b22ff0097171fa0f8c6b1

Reply:
because main() is 0x00000000402570  but you call into 0x7fffffffeb10?

Reply:
I'd guess 0x7fffffffeb10 is the address of stack local var main_hook

Reply:
Thanks for the nudge.
I have fixed the pointer. So the rax address should be pointed to the right address. But it resulted in SIGABRT.

```
[ Legend: Modified register | Code | Heap | Stack | String ]
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── registers ────
$rax   : 0x00000000402570  →  <main+0> push rbp
$rbx   : 0x0               
$rcx   : 0x100             
$rdx   : 0x007fffffffeb78  →  0x007fffffffedba  →  "HOSTNAME=157cb61fb3d1"
$rsp   : 0x007fffffffea40  →  0x007fffffffeb68  →  0x007fffffffed78  →  "/workspaces/LibAFL/fuzzers/frida_executable_libpng[...]"
$rbp   : 0x000000004246f0  →  <__libc_csu_init+0> push r15
$rsi   : 0x007fffffffeb68  →  0x007fffffffed78  →  "/workspaces/LibAFL/fuzzers/frida_executable_libpng[...]"
$rdi   : 0x1               
$rip   : 0x007ffff5763d08  →  <__libc_start_main+232> call rax
$r8    : 0x0               
$r9    : 0x007ffff7fe21b0  →  <_dl_fini+0> push rbp
$r10   : 0x007ffff7fef180  →  <strcmp+3824> pxor xmm0, xmm0
$r11   : 0xa               
$r12   : 0x000000004246f0  →  <__libc_csu_init+0> push r15
$r13   : 0x0               
$r14   : 0x007ffff7fe21b0  →  <_dl_fini+0> push rbp
$r15   : 0x00000000424750  →  <__libc_csu_fini+0> ret 
$eflags: [ZERO carry PARITY adjust sign trap INTERRUPT direction overflow resume virtualx86 identification]
$cs: 0x33 $ss: 0x2b $ds: 0x00 $es: 0x00 $fs: 0x00 $gs: 0x00 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── stack ────
0x007fffffffea40│+0x0000: 0x007fffffffeb68  →  0x007fffffffed78  →  "/workspaces/LibAFL/fuzzers/frida_executable_libpng[...]"	 ← $rsp
0x007fffffffea48│+0x0008: 0x00000001c3afd500
0x007fffffffea50│+0x0010: 0x00000000402570  →  <main+0> push rbp
0x007fffffffea58│+0x0018: 0x007ffff5876b4f  →  <_dl_catch_error+47> mov rdx, QWORD PTR [rsp]
0x007fffffffea60│+0x0020: 0x0000000000000000
0x007fffffffea68│+0x0028: 0xf6d68178fea7352c
0x007fffffffea70│+0x0030: 0x000000004246f0  →  <__libc_csu_init+0> push r15
0x007fffffffea78│+0x0038: 0x0000000000000000
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x7ffff5763cfc <__libc_start_main+220> mov    rsi, QWORD PTR [rsp]
   0x7ffff5763d00 <__libc_start_main+224> mov    rdx, QWORD PTR [rax]
   0x7ffff5763d03 <__libc_start_main+227> mov    rax, QWORD PTR [rsp+0x10]
 → 0x7ffff5763d08 <__libc_start_main+232> call   rax
   0x7ffff5763d0a <__libc_start_main+234> mov    edi, eax
   0x7ffff5763d0c <__libc_start_main+236> call   0x7ffff577b660 <__GI_exit>
   0x7ffff5763d11 <__libc_start_main+241> mov    rax, QWORD PTR [rsp]
   0x7ffff5763d15 <__libc_start_main+245> lea    rdi, [rip+0x171d0c]        # 0x7ffff58d5a28
   0x7ffff5763d1c <__libc_start_main+252> mov    rsi, QWORD PTR [rax]
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────── arguments (guessed) ────
*0x402570 (
   $rdi = 0x00000000000001,
   $rsi = 0x007fffffffeb68 → 0x007fffffffed78 → "/workspaces/LibAFL/fuzzers/frida_executable_libpng[...]",
   $rdx = 0x007fffffffeb78 → 0x007fffffffedba → "HOSTNAME=157cb61fb3d1",
   $rcx = 0x00000000000100
)
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "libpng-harness", stopped 0x7ffff5763d08 in __libc_start_main (), reason: SINGLE STEP
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x7ffff5763d08 → __libc_start_main(main=0x402570 <main(int, char**)>, argc=0x1, argv=0x7fffffffeb68, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffeb58)
[#1] 0x7ffff5ca63c3 → frida_fuzzer::__libc_start_main(main=0x402570 <main(int, char**)>, argc=0x1, argv=0x7fffffffeb68, init=0x4246f0 <__libc_csu_init>, fini=0x424750 <__libc_csu_fini>, rtld_fini=0x7ffff7fe21b0 <_dl_fini>, stack_end=0x7fffffffeb58)
[#2] 0x40237a → _start()
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  ni

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=0x6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
```

Reply:
Thanks for the nudge again. This issue is fixed. I could call rust main_hook function by using original __libc_start_main function.